### DISCUSSION TITLE
Add a way to debounce signal handler

### DISCUSSION THREAD

**Discussion by timothyqiu** [godotengine/godot-proposals#11267] [Engine Core]
Inspired by https://github.com/godotengine/godot/pull/99812, I believe debouncing is a common requirement for controls that accept user input.

We can provide a special `Callable` that calls the wrapped child `Callable` in a debounced manner.

```gdscript
# Callable.debounced(seconds: float) -> DebouncedCallable
line_edit.text_changed.connect(_some_handler.debounced(0.5))
```

I am not familiar with the implementation of `Callable`. How to obtain the time may require further design.

**Comment by sockeye-d** [comment]
I made this:
```gdscript
func debounced(fn: Callable, debounce_sec: float) -> Callable:
	# is array because they are passed by reference
	# (it's hacky, I know)
	var last_call_usec: PackedInt64Array = [-1.0]
	return func():
		if last_call_usec[0] == -1.0 or (Time.get_ticks_usec() - last_call_usec[0]) * 1e-6 > debounce_sec:
			last_call_usec[0] = Time.get_ticks_usec()
			fn.call()
```
It's kind of goofy but it seems to work

**Reply by Iaotle** [reply]
I needed to use this as a preload hook, this is what I ended up with:
```gdscript
var last_call_usec: PackedInt64Array = [-1.0]

func debounced(fn: Callable, debounce_sec: float, params) -> Callable:
	return func():
		if last_call_usec[0] == -1.0 or (Time.get_ticks_usec() - last_call_usec[0]) * 1e-6 > debounce_sec:
			last_call_usec[0] = Time.get_ticks_usec()
			if (params):
				fn.call(params)
			else:
				fn.call()
```
Now I just call: 
```gdscript
Debounce.debounced(function, DELAY, params).call_deferred()
```

**Reply by sockeye-d** [reply]
You don't actually need to pass params into it, you can just bind them or wrap the function in a lambda:
```gdscript
Debounce.debounced(function.bind("param1", "param2"), DELAY)
# or
Debounce.debounced(func(): function("param1", "param2"), DELAY)
```

**Comment by Calinou** [comment]
This could also be provided for throttling, but debouncing is arguably needed more often in UI interactions.

**Comment by adamscott** [comment]
@dalexeev What do you think? I think you're one of the most knowledgeable person I know about `Callable`, if I'm not mistaken.

**Reply by dalexeev** [reply]
Technically this can be done with a `CustomCallable`. The question is whether we should support it out of the box, given that users can implement it themselves with `SceneTreeTimer`/`Tween`. Also, how much can the required behavior vary? Is there a single option that will satisfy all users?

**Reply by timothyqiu** [reply]
@dalexeev The point of this proposal is indeed to reduce the user's boilerplate code.

ReactiveX established [an API](https://reactivex.io/documentation/operators/debounce.html) for debouncing that seems to satisfy all users.

**Reply by adamscott** [reply]
I think it could greatly improve UX inside the editor as well.

**Comment by passivestar** [comment]
If godot provides signal debouncing/throttling on the engine level shouldn't it be done in a way that's compatible with persisting connections so that it can be serialized and edited through the editor UI, just like one-shot and deferred?

**Comment by NoctemCat** [comment]
Here is another hacky implementation:
```gdscript
@tool
@abstract
class_name Utils


static func debounce(callable: Callable, delay: float = 0.2) -> Callable:
    var state: Array
    return _debounce_call.bind(callable, state, delay)

static func _debounce_call(callable: Callable, state: Array, delay: float) -> void:
    if state.size() > 0:
        state[0].kill()
        state.clear()
        
    var tween := _get_tween()
    tween.tween_callback(callable).set_delay(delay)
    tween.tween_callback(state.clear)
    state.push_back(tween)


static func throttle(callable: Callable, limit: float = 0.2) -> Callable:
    var state: Array
    return _throttle_call.bind(callable, state, limit)

static func _throttle_call(callable: Callable, state: Array, limit: float) -> void:
    if state.size() > 0:
        state[1] = true
        return

    var tween := _get_tween()
    tween.tween_callback(_throttle_trail_check.bind(callable, state, limit)).set_delay(limit)
    state.push_back(tween)
    state.push_back(false)
    callable.call()

static func _throttle_trail_check(callable: Callable, state: Array, limit: float) -> void:
    if state[1]: 
        state.clear()
        _throttle_call(callable, state, limit)
    else:
        state.clear()


static func _get_tween() -> Tween:
    var tree: SceneTree = Engine.get_main_loop()
    var tween := tree.create_tween()
    tween.set_pause_mode(Tween.TWEEN_PAUSE_PROCESS)
    tween.set_ignore_time_scale(true)
    return tween
```
It's also super goofy.

It should work like a similarly named things used in Javascript.

**Comment by sockeye-d** [comment]
I'm *pretty* sure this works too:
```gdscript
# register this as an autoload
extends Node

func debounce(timeout_seconds: float, what: Callable) -> Callable:
    var last_timer: Array[SceneTreeTimer] = [null]
    return func(...args):
        if last_timer[0]:
            last_timer[0].free()
        last_timer[0] = get_tree().create_timer(timeout_seconds)
        last_timer[0].timeout.connect(func(): return what.callv(args))
```

**Reply by NoctemCat** [reply]
The idea makes sense, but I don't think you can call `free` on `RefCounted`

**Reply by sockeye-d** [reply]
It seems to work fine. It's probably not recommended, if you really wanted to force-free a RefCounted you could also do

```gdscript
while last_timer[0].get_reference_count() > 0:
    last_timer[0].unreference
```

**Reply by NoctemCat** [reply]
<img width="638" height="176" alt="image" src="https://github.com/user-attachments/assets/e2900d8c-c0cd-4c03-85a5-af94488984fa" />
Doesn't look fine to me.

And unreference version crashed on me.

**Reply by sockeye-d** [reply]
I wrote it wrong, unreference is supposed to be called as a method like
```gdscript
last_timer[0].unreference()
```

**Reply by NoctemCat** [reply]
I called it as a method, yes, and it crashed when using it

**Reply by sockeye-d** [reply]
Odd. Well, I guess you'd have to store a reference to the last connected Callable and manually disconnect it:
```gdscript
func debounce(timeout_seconds: float, what: Callable) -> Callable:
    var last_timer: Array[SceneTreeTimer] = [null]
    var last_callable: Array[Callable] = [Callable()]
    return func(...args):
        if last_timer[0] and last_callable[0]:
            last_timer[0].timeout.disconnect(last_callable[0])
        last_timer[0] = get_tree().create_timer(timeout_seconds)
        last_callable[0] = func(): return what.callv(args)
        last_timer[0].timeout.connect(last_callable[0])
```
